controller/contour/daemonset.go

```diff
@@ -0,0 +1,395 @@
+	// envoyCfgVolName is the name of the Envoy configuration volume.
+	envoyCfgVolName = "envoy-config"
+	// envoyCfgVolMntDir is the directory name of the Envoy configuration volume.
+	envoyCfgVolMntDir = "config"
+	// envoyCfgFileName is the name of the Envoy configuration file.
+	envoyCfgFileName = "envoy.json"
+	// envoyDaemonSetLabel identifies a daemonset as a contour daemonset,
+	// and the value is the name of the owning contour.
+	envoyDaemonSetLabel = "contour.operator.projectcontour.io/daemonset-envoy"
+)
+
+// ensureDaemonSet ensures a DaemonSet exists for the given contour.
+func (r *Reconciler) ensureDaemonSet(ctx context.Context, contour *operatorv1alpha1.Contour) error {
+	desired := DesiredDaemonSet(contour, r.Config.ContourImage, r.Config.EnvoyImage)
+
+	current, err := r.currentDaemonSet(ctx, contour)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			if err := r.createDaemonSet(ctx, desired); err != nil {
+				return fmt.Errorf("failed to create daemonset for contour %s/%s: %w", contour.Namespace, contour.Name, err)
+			}
+			return nil
+		}
+		return fmt.Errorf("failed to get daemonset for contour %s/%s: %w", contour.Namespace, contour.Name, err)
+	}
+
+	if err := r.updateDaemonSetIfNeeded(ctx, current, desired); err != nil {
+		return fmt.Errorf("failed to update daemonset for contour %s/%s: %w", contour.Namespace, contour.Name, err)
+	}
+
+	return nil
+}
+
+// ensureDaemonSetDeleted ensures the DaemonSet for the provided contour is deleted.
+func (r *Reconciler) ensureDaemonSetDeleted(ctx context.Context, contour *operatorv1alpha1.Contour) error {
+	ds := &appsv1.DaemonSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: contour.Spec.Namespace.Name,
+			Name:      contour.Name,
+		},
+	}
+
+	if err := r.Client.Delete(ctx, ds); err != nil {
+		if errors.IsNotFound(err) {
+			return nil
+		}
+		return fmt.Errorf("failed to delete daemonset for contour %s/%s: %w", contour.Namespace, contour.Name, err)
+	}
+
+	r.Log.Info("deleted daemonset", "namespace", ds.Namespace, "name", ds.Name)
+	return nil
+}
+
+// DesiredDaemonSet returns the desired DaemonSet for the provided contour using
+// contourImage as the shutdown-manager/envoy-initconfig container images and
+// envoyImage as Envoy's container image.
+func DesiredDaemonSet(contour *operatorv1alpha1.Contour, contourImage, envoyImage string) *appsv1.DaemonSet {
+	parsedImage := strings.Split(contourImage, ":")
+	imageTag := parsedImage[1]
+	labels := map[string]string{
+		"app.kubernetes.io/name":     "contour",
+		"app.kubernetes.io/instance": contour.Name,
+		// The contourImage tag is used as the version.
+		"app.kubernetes.io/version":    imageTag,
+		"app.kubernetes.io/component":  "ingress-controller",
+		"app.kubernetes.io/managed-by": "contour-operator",
+		// Associate the daemonset with the provided contour.
+		owningContourLabel: contour.Name,
+	}
+
+	pointerTo := func(ios intstr.IntOrString) *intstr.IntOrString { return &ios }
```

Comment:
You might as well do this:
```Go
pointerTo := func(i int) *intstr.IntOrString {
  val := intstr.IntOrString{IntVal: int32(i)}
  return &val
}
```

Reply:
With your suggested changes:
```
$ make test
/Users/daneyonhansen/code/go/bin/controller-gen object:headerFile="hack/boilerplate.go.txt" paths="./..."
go fmt ./...
go vet ./...
# github.com/projectcontour/contour-operator/controller/contour
controller/contour/controller.go:35:41: cannot convert i (type intstr.IntOrString) to type int32
# github.com/projectcontour/contour-operator/controller/contour
vet: controller/contour/controller.go:35:42: cannot convert i (variable of type intstr.IntOrString) to int32
make: *** [vet] Error 2
```

I did move the func up in scope to remove duplication with `deployment.go`.

Reply:
Well clearly you'd want to update the call sites too 😂